File: AI/ml/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/MLR_MASTER.py
# Data Preprocessing MLR

##################################
### check code     : cmd + i   ###
### run code       : cmd enter ###
### set working dir: Fn F5     ###
##################################

# Import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Import dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Split dataset into Training and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Categorical Value: State
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()
X

# NOTE: Categorical Value: Avoid Dummy Var trap (Python lib usually takes care of) 
X=X[:,1:]

# Split dataset into Training and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Feature Scaling (optional) necessary if there are Euclidian distance to be processed
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

# Fit MLR to ***Training*** set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predict the ***Test*** set results, aka vector of prediction
y_pred = regressor.predict(X_test)
y_pred

# Build the OPTIMAL model using *** Backward Elimination ***
import statsmodels.formula.api as sm
# API does not care for X0=1,  so at beginning add
#                      "1's  50  1                                  col"
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)

# The OPTIMAL, prevailing matrix
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
# Look for predictor P-value for each var, compare it with SL 0.05, and remove
regressor_OLS.summary()
# Remove predictor x2 (p-val 0.99) of prevailing matrix
X_opt = X[:, [0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# Remove predictor x1 of prevailing matrix
X_opt = X[:, [0,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# Remove predictor x2 of prevailing matrix
X_opt = X[:, [0,3,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# Remove predictor x2 of prevailing matrix, and what remains is R&D col
X_opt = X[:, [0,3]] 
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
# # # # # # # # # # # # # # # # # # # # 

#def backwardElimination(x, sl):
#    numVars = len(x[0])
#    for i in range(0, numVars):
#        regressor_OLS = sm.OLS(y, x).fit()
#        maxVar = max(regressor_OLS.pvalues).astype(float)
#        if maxVar > sl:
#            for j in range(0, numVars - 1):
#                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
#                    x = np.delete(x, j, 1)
#    regressor_OLS.summary()
#   return x
 
#SL = 0.05
#X_opt = X[:, [0, 1, 2, 3, 4, 5]]
#X_Modeled = backwardElimination(X_opt, SL)

File: AI/ml/Machine Learning A-Z Template Folder/Part 2 - Regression/Section 5 - Multiple Linear Regression/MLR_MASTER.R
# Data Preprocessing MLR

############### R ####################
### check code      : Fn F1        ###
### run code        : cmd enter    ###
### set working dir : files 'more' ###
######################################

# Importing the dataset
dataset = read.csv('50_Startups.csv')
# dataset = dataset[, 2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Encoding categorical data using factor function (c being a vector)
dataset$State = factor(dataset$State,
                         levels = c('New York', 'California', 'Florida'), 
                         labels = c(1,2,3))

# Splitting the dataset into the Training set and Test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

# Feature Scaling (optional) necessary if there are Euclidian distance to be processed
# Here it will be take care of by fit function


# Fitting Multiple Linear Regression to Training set
regressor = lm(formula = Profit ~ ., 
               data = training_set)
# NOTE: R builds and if necessary removes one Dummy Vars
summary(regressor)

# Predicting the Test set results (10) 
y_pred = predict(regressor, newdata = test_set)
y_pred

# BACKWARD ELIMINATION == Find vars that have effect
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State, 
               data = dataset) #NOTE: the training_set
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend, 
               data = dataset) 
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend, 
               data = dataset) 
summary(regressor)

regressor = lm(formula = Profit ~ R.D.Spend, 
               data = dataset) #NOTE Marketing had moved from P=10 to P=6
summary(regressor)

















